### PR TITLE
Integrate TTML and Add AdamW Op

### PR DESCRIPTION
**Pull request by pglusacTT**
### Ticket
N/A

### Problem description
We need TTML-specific ops for training.

### What's changed
Wires tt-train's fused AdamW step into tt-mlir end to end: TTIR and TTNN ops, the StableHLO composite path, TTIR->TTNN and EmitC conversions, a flatbuffer schema, and the runtime handler that calls `ttml::metal::adamw` on device. Golden, ttir_builder, and lit/python tests included.

### Notes
Build options considered:

- Turn on `BUILD_TT_TRAIN`, link `libttml.a`. Rejected: It causes the ABI mismatch
  for the Python bindings and pulls in a second ttnn.
- Ship the ops as a separate satellite `.so`. Rejected: Extra packaging work, illogical separation.
- Patch tt-train's CMake to skip its nanobind or build a shared lib. Rejected: Fragile but not a precedented.
- Compile the ttml op sources ourselves against the shared ttnn. Chosen: We
  already have the sources in the checkout, so tt-train never needs building.
  `BUILD_TT_TRAIN` stays off (no ABI bug), there's one ttnn, no satellite, no patch.

The chosen path is a small CMake target (`third_party/ttml`). We compile only the ops the runtime
dispatches (AdamW), since only referenced ops get linked and many need `ttml::core` we don't build. It's a static lib so unreferenced ops with unresolved symbols are tolerated and it stays out of the runtime's NEEDED list. 

Runtime dispatch follows the normal pattern: `program_executor` calls `operations::ttml::run(op->type_as_AdamWOp(), context)` directly

EmitPy lowering is intentionally unsupported and reports a clean match failure: it would need the low-level `ttml::metal::adamw` primitive, but tt-train's bindings only expose the high-level optimizer class. EmitC is supported.

### CODE REVIEW

**Comment by vladimirjovanovicTT** [comment]
@sdjordjevicTT can you please have a look? :)

**Review by ndrakulicTT — APPROVED** [review]
Golden LGTM

File: third_party/CMakeLists.txt [hunk 1]
```diff
@@ -247,6 +251,15 @@ install(
   REGEX "third_party/umd/"
 )
 
+install(
+  DIRECTORY
+    ${TTML_SOURCE_DIR}/metal
+  DESTINATION ${CMAKE_INSTALL_PREFIX}/tt-metal/${TTML_RELPATH}
+  USE_SOURCE_PERMISSIONS
+  COMPONENT SharedLib
+  EXCLUDE_FROM_ALL
```
Comment (mmilosevicTT) [inline third_party/CMakeLists.txt:260]: Why does this differ from `umd` install above?

File: CMakeLists.txt [hunk 2]
```diff
@@ -171,6 +171,7 @@ add_definitions(${LLVM_DEFINITIONS})
 include(TTMLIRPythonSitePackages)
 
 add_subdirectory(third_party)
+add_subdirectory(third_party/ttml)
```
Comment (mmilosevicTT) [inline CMakeLists.txt]: do we really need this here? I would expect setup for ttml to completely be in the cmake from third_party and ttml.

File: tools/golden/mapping.py [hunk 3]
```diff
@@ -9046,6 +9106,7 @@ def debug_region_end_golden(
     ttnn.LayerNormPostAllGatherOp: ttnn_layer_norm_post_all_gather_golden,
     ttnn.GroupNormOp: ttnn_group_norm_golden,
     ttnn.RMSNormOp: rms_norm_golden,
+    ttnn.AdamWOp: adamw_golden,
```
Comment (jgrimTT) [inline tools/golden/mapping.py]: Nit: remove unused ttnn mapping

File: test/python/golden/test_ttir_ops.py [hunk 4]
```diff
@@ -172,6 +172,48 @@ def div(
...
+            grad: Operand,
+            exp_avg: Operand,
+            exp_avg_sq: Operand,
+            builder: TTIRBuilder,
+            unit_attrs: Optional[List[str]] = None,
+        ):
+            exp_avg_sq_t = builder._get_golden_tensor(exp_avg_sq).apply_shardwise(
+                lambda s: s.abs()
+            )
+            builder.set_goldens_from_builder_tensor({exp_avg_sq: exp_avg_sq_t}, {})
```
Comment (jgrimTT) [inline test/python/golden/test_ttir_ops.py:194]: Just curious, what's the purpose of the abs() living outside the golden mapping?

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td
Comment (djstefanovicTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td:1]: Do we need a separate TTIR op for this? Conversion from TTIR to TTNN is linear. Can it be done with TTIR composites? @sdjordjevicTT @pglusacTT what do you think?

File: lib/Conversion/TTNNToEmitPy/TTNNToEmitPy.cpp
Comment (djstefanovicTT) [inline lib/Conversion/TTNNToEmitPy/TTNNToEmitPy.cpp:1]: Can you make issue to link here instead of TODO?

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td
Comment (sdjordjevicTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td:1]: If this is a fused op, then yeah, a composite is a better fit.

File: third_party/CMakeLists.txt (hunk 1 above)
Comment (pglusacTT) [reply third_party/CMakeLists.txt:260]: It matches the `runtime/` and `ttnn/` installs, don't know what happened to umd.

File: test/python/golden/test_ttir_ops.py (hunk 4 above)
Comment (pglusacTT) [reply test/python/golden/test_ttir_ops.py:194]: It's a part of the input setup, the second moment is always non-negative.

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td
Comment (pglusacTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td:1]: Synced offline; will proceed with a dedicated TTIR op for now.

**Review by djstefanovicTT — APPROVED** [review]

**Review by jgrimTT — APPROVED** [review]
Builder LGTM

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 5]
```diff
@@ -2338,6 +2338,51 @@ def TTIR_RMSNormOp : TTIR_NamedOp<"rms_norm", [AttrSizedOperandSegments]> {
...
+    - `beta1_pow`, `beta2_pow`: beta1^t and beta2^t bias-correction terms.
+    - `epsilon`: numerical-stability constant.
+    - `weight_decay`: decoupled weight-decay coefficient.
+    - `stochastic_rounding`: enable stochastic rounding (default false).
+
+    Output:
+    - `result` (Tensor): the updated parameter tensor.
+  }];
+
+  let arguments = (ins Arg<AnyRankedTensor, "parameter tensor (updated in place)", [MemWrite]>:$param,
```
Comment (amalkovTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td]: Just curious, why is the updated parameter tensor both written in place and returned as a result?

**Review by acicovicTT — APPROVED** [review]

**Review by anuragsingh-tt — APPROVED** [review]

**Review by acolicTT — APPROVED** [review]

**Review by nsmithtt — APPROVED** [review]

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td (hunk 5 above)
Comment (pglusacTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td]: Thanks for spotting this. TTML has a weird interface where it does inplace update but still returns that tensor, I missed that. I remodeled the OP so that it models this in a better way. It is similar to Batch Norm Training OP now. Please check https://github.com/tenstorrent/tt-mlir/pull/9044/changes/47d1403e8abd06bfa55001b7baa89dc307911def.

**Review by amalkovTT — APPROVED** [review]
